packages/lodestar/src/chain/beaconProposerCache.ts

```diff
@@ -0,0 +1,37 @@
+import {MapDef} from "../util/map";
+import {Epoch} from "@chainsafe/lodestar-types";
+import {IMetrics} from "../metrics";
+
+const PROPOSER_PRESERVE_EPOCHS = 2;
+export type ProposerPreparationData = {
+  validatorIndex: string;
+  feeRecipient: string;
+};
+
+export class BeaconProposerCache {
+  private readonly feeRecipientByValidatorIndex: MapDef<string, {epoch: Epoch; feeRecipient: string}>;
+  constructor(opts: {defaultFeeRecipient: string}, private readonly metrics?: IMetrics | null) {
+    this.feeRecipientByValidatorIndex = new MapDef<string, {epoch: Epoch; feeRecipient: string}>(() => ({
+      epoch: 0,
+      feeRecipient: opts.defaultFeeRecipient,
+    }));
+  }
+
+  add(epoch: Epoch, {validatorIndex, feeRecipient}: ProposerPreparationData): void {
+    this.feeRecipientByValidatorIndex.set(validatorIndex, {epoch, feeRecipient});
+  }
+
+  prune(epoch: Epoch): void {
```

Comment:
normally we use epoch as the key and the `prune()` is much simpler, why do we use `validatorIndex` as the key here?

Reply:
nvm, we use `validatorIndex` as key because validator only provide feeRecipient for 1 epoch but we could use it for the next 2 epochs